bip-xxxx.mediawiki

```diff
@@ -0,0 +1,358 @@
+* If the sender's inputs are all from the same scriptPubKey type, the receiver must match the same type. If the receiver can't match the type, they must return error <code>unavailable</code>.
+* <code>maxadditionalfeecontribution=</code> and <code>minfeerate=</code> should be ignored in the [[#spare-change|spare change]] case.
+
+<code>*</code>: Interactive receivers are not required to validate the original PSBT because they are not exposed to probing attacks.
+
+===Sender's payjoin proposal checklist===
+
+The sender should check the payjoin proposal before signing it to prevent a malicious receiver from stealing money.
+
+* Check that all the spent outpoints in the original PSBT still exist in the coinjoin PSBT.
+* Check that all the spent outpoints in the original PSBT do not have any partial signature.
+* If the sender is not using inputs with mixed types, check that the receiver inputs type match the inputs type of the sender. (ie. both using P2SH-P2WPKH or both using P2WPKH)
+* Check that any inputs added by the receiver are finalized.
+* Check that the transaction version, and nLockTime are unchanged.
+* Check that the sender's inputs' sequence numbers are unchanged.
+* If the sender's inputs' sequence numbers are all the same, check that the receiver's contributed inputs match those.
+* Check that the sender's own outputs have not been modified, except for paying increased fee, or by simple shuffling.
+* If sender specified <code>additionalfeeoutputindex=</code> (see later), the fee should have been subtracted from the output at the same index in the original PSBT.
+* If <code>minfeerate</code> was specified, check that the estimated fee rate of the payjoin proposal is not less than this value.
+* Check that the sent amount in the payjoin proposal is less than or equal to the sent amount of the original transaction. (Defined as the sum of the inputs' value to be signed minus the sender's ouput change)
+
+If the sent amount in the payjoin proposal is above the amount sent in the original PSBT
+* Check that the additional paid amount has been add paid to the fee.
+* Check that the estimated fee rate of the payjoin proposal is not more than the fee rate of the original PSBT. (fee estimation is hard, so we should allow ~2 satoshi per inputs as margin of error)
+* If <code>maxadditionalfeecontribution=</code> was specified, check the additional paid amount is less than or equal to this amount.
+* If <code>maxadditionalfeecontribution=</code> was not specified, the sender's software should ask an interactive confirmation to the user.
+* <code>maxadditionalfeecontribution=</code> and <code>minfeerate=</code> should be ignored in the [[#spare-change|spare change]] case.
+
+The sender must be careful to only sign the inputs that were present in the original PSBT and nothing else.
+
+Note:
+* The sender should allow the payment output to be modified by the receiver (The receiver may substitute a P2WPKH payment to P2SH payment to increase privacy)
```

Comment:
I am not at all thrilled about the possibility of replacing outputs, it seems to create too many complications. In any case, I think several clarifications are needed here:
1. If the recipient modifies the output's script type, do the two scripts have to be using the same public key? If yes, then that might be problematic for some wallets in terms of account discovery, because the script type is often linked to the BIP32 path (m/49' for P2SH-P2WPKH and m/84' for P2WPKH). If no, then I can imagine some UX problems. Consider for example a sender who verifies the recipient's address using a second channel before signing the original transaction. Afterwards, the recipient's address is modified in the CoinJoin proposal, so the sender needs to go through another round of address verification.
2. What do the words "should allow" exactly mean? To me it says that the recipient can't rely on the sender allowing output replacement, so it's probably better not to even try it unless absolutely necessary. I think it should be either formulated as "must allow" or preferably, there should be a HTTP query string parameter telling the recipient whether the sender allows output replacement or not.

Reply:
the sender can just shoot off the original as soon as it gets a proposal it doesn't like if it wants. Is that not enough?

Reply:
I assume you are referring to point 2. The sender certainly can shoot off the original, but it's a wasted opportunity. The recipient attempts to improve an already good CoinJoin proposal by replacing one of it's outputs, but by doing so it causes it to fail and revert to a standard transaction.

Reply:
I also think replacing outputs is a bad idea.

One really nice thing about BIP79 imho is that a secure connection to the receiver is not critical. Even if the "payjoin" server (or connection) is compromised, it is just limited to causing privacy loss. If replacing outputs is allowed, the funds can just be redirected wholesale. From a practical point of view, this can make it a lot more difficult to deploy (or outsource)

Reply:
I think you are right. Actually we could activate it later via an additional optional parameter.

Reply:
> If the recipient modifies the output's script type, do the two scripts (the original and the modified) have to be using the same public ke

This is up to the receiver to handle that, this should not be part of the BIP how he achieves this. (That said, we don't support it yet in btcpayserver)

> so the sender needs to go through another round of address verification.

Well, in any case, the sender need to do two verifications. Especially with hardware wallet.

Reply:
> Well, in any case, the sender need to do two verifications. Especially with hardware wallet.

Not necessarily. If some basic rules are respected, then the hardware wallet can be given the original signed transaction, verify the signatures and verify that the PayJoin proposal only extends the original transaction without modifying the original destination address or the amount that the user agreed to spend. Thus it can safely re-sign the inputs with minimal or no burden to the user. The basic rules that I am referring to are effectively those formulated in the [Output Adjustment](https://github.com/bitcoin/bips/blob/master/bip-0079.mediawiki#output-adjustment) section of BIP-0079: The receiver MUST NOT remove any inputs or outputs and MUST NOT decrease any output amount.

Reply:
> > If the recipient modifies the output's script type, do the two scripts (the original and the modified) have to be using the same public key?

> This is up to the receiver to handle that, this should not be part of the BIP how he achieves this. (That said, we don't support it yet in btcpayserver)

One of the reasons why I ask about this is that if both script types could use the same public key, then we could leave the choice about which one to use to the sender, which means that the receiver wouldn't need to modify the outputs. The receiver would always provide a P2WPKH and the sender would be free to change it to P2SH if the sender's inputs are P2SH. On the other hand, some recipients might not be happy about the sender changing the script type due to the reasons I explained in the first comment. The recipient could indicate his policy on this in the BIP-0021 URI. Another possibility is for the sender to provide a set of addresses in the URI, each using a different type, so that the sender can choose between them. Although BIP-0021 wasn't really built for that, it could be done by adding an "alt-address" parameter.